packages/salesforcedx-vscode-core/src/commands/forceApexTestRunCodeAction.ts

```diff
@@ -10,13 +10,51 @@ import {
   SfdxCommandBuilder
 } from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
 import { nls } from '../messages';
+import { notificationService } from '../notifications';
 import {
   EmptyParametersGatherer,
   SfdxCommandlet,
   SfdxCommandletExecutor,
   SfdxWorkspaceChecker
 } from './commands';
 
+function isEmpty(value: string): boolean {
+  return !value || value.length === 0;
+}
+
+function isNotEmpty(value: string): boolean {
+  return !isEmpty(value);
+}
+
+// cache last test class and test method values to
+// enable re-running w/o command context via built-in LRU
+class ForceApexTestRunCache {
```

Comment:
Convention: Please call this `ForceApexTestRunCacheService`. The convention is to use the word `Service` to denote any singletons.

Reply:
Will do.